### PR TITLE
[WIP] clean some "for range(len(foo))" usage

### PR DESCRIPTION
**Pull request by mwichmann**
NOTE: this is being pushed to have a PR to try experiments on the `Interactive/configure.py` test. There's nothing wrong with the subject of the PR as such, but it's not high prio at all - I'll update this summary if the state changes.  The configure test does use one of the match functions which got some updates, though the current updates should not affect the test behavior. Making it a draft since apparently appveyor builds still trigger on drafts, and that'e where we are seeing the fail.

====
Looping over a range given by the length of an object is no longer the preferred idiom; instead just loop over the object, using `enumerate` if we *really* need the index as well.

Along the way, document some test functions, and update the docs on some other test functions.

Signed-off-by: Mats Wichmann <mats@linux.com>

## Contributor Checklist:

* [ ] I have created a new test or updated the unit tests to cover the new/changed functionality.
* [ ] I have updated `master/src/CHANGES.txt` directory (and read the `README.txt` in that directory)
* [ ] I have updated the appropriate documentation

### CODE REVIEW

**Comment by mwichmann** [comment]
How helpful... create a PR to chase an error, and the error doesn't occur during the build?

File: testing/framework/TestCmdTests.py [hunk 1]
```diff
@@ -424,10 +430,13 @@ class diff_TestCase(TestCmdTestCase):
     def test_diff_re(self):
         """Test diff_re()"""
         result = TestCmd.diff_re(["abcde"], ["abcde"])
+        result = list(result)
```
Comment (bdbaddog) [inline testing/framework/TestCmdTests.py:433]: Why is this necessary? Doesn't diff_re() still return a list?

File: testing/framework/TestCmdTests.py [hunk 2]
```diff
@@ -1059,16 +1070,16 @@ def test_evaluation(self):
 class match_re_dotall_TestCase(TestCmdTestCase):
     def test_match_re_dotall_function(self):
         """Test calling the TestCmd.match_re_dotall() function"""
-        assert TestCmd.match_re_dotall("abcde\nfghij\n", "a.*j\n")
+        assert TestCmd.match_re_dotall("abcde\nfghij\n", r"a.*j\n")
```
Comment (bdbaddog) [inline testing/framework/TestCmdTests.py:1073]: Doesn't this break the \n ?

**Comment by bdbaddog** [comment]
This is a bigger PR to review. It'd be a bit easier if you could break it up into a few smaller sets of files.
(The diff is many pages of changes to review, especially given these are all cleanup type changes and don't need to be committed together for the most part)

**Comment by mwichmann** [comment]
That's why it's in draft state... I'm using a batch of pending changes lumped together to check out that one test fail - which doesn't seem to happen now.  When there are things ready for review, they'll be dribbled out in a more digestible fashion (and this one withdrawn), don't worry.

File: testing/framework/TestCmdTests.py (hunk 1 above)
Comment (mwichmann) [reply testing/framework/TestCmdTests.py:433]: yes.  I have thoughts of diff_re and simple_diff becoming generators like the difflib functions they shadow - it's kind of odd to have to handle the returns of some diff functions one way and some another way.  Don't know if I'll ever push that, though.

**Comment by mwichmann** [comment]
I found what I needed from this one, so withdrawing.  A series of smaller PRs covering the details are on their way.

**Comment by mwichmann** [comment]
several pieces have been submtted individually, dropping this one
